### ISSUE TITLE
Extern namespaces seem to be wrong

### ISSUE THREAD

**Issue by tadeegan** [angular/tsickle#369]
Not sure if this is covered by another open issue however I am seeing classes declared within a namespace have their namespace dropped in the output extern.js.

```
declare namespace React {
    type Key = string | number;
    
    interface ReactElement<P> {
        type: string | ComponentClass<P>;
        props: P;
        key: Key | null;
    }

    interface ComponentClass<P> {
        defaultProps?: Partial<P>;
        displayName?: string;
    }
}
```

Compiles to:

```
// externs from test.d.ts:
/**
 * @externs
 * @suppress {duplicate}
 */
// NOTE: generated by tsickle, do not edit.
/** @const */
var React = {};

/** @typedef {(string|number)} */
React.Key;

/** @record @struct */
React.ReactElement = function() {};
 /** @type {(string|!ComponentClass<?>)} */
React.ReactElement.prototype.type;
 /** @type {?} */
React.ReactElement.prototype.props;
 /** @type {(string|number)} */
React.ReactElement.prototype.key;

/** @record @struct */
React.ComponentClass = function() {};
 /** @type {?} */
React.ComponentClass.prototype.defaultProps;
 /** @type {string} */
React.ComponentClass.prototype.displayName;

```

Note that "ComponentClass" is referred to without the "React" namespace. Closure compiler chokes on this.

**Comment by evmar** [comment]
Does it work with the `--untyped` flag?

**Comment by tadeegan** [comment]
I mean, yes it works in these sense that all types are `{*} `now but that defeats the whole purpose of this tool...

**Comment by evmar** [comment]
externs are required to prevent renaming, so even {*} is providing something valuable -- without these externs your code doesn't compile, and with them it does.

TypeScript has already type-checked your code, so having Closure compiler do more checks is not very valuable.

**Comment by tadeegan** [comment]
> TypeScript has already type-checked your code, so having Closure compiler do more checks is not very valuable.

Not true. The closure compiler uses the type information to perform type based minfication. To fully take advantage of that you need well typed externs.  

Should we just set all the types in:
https://github.com/google/closure-compiler/tree/master/contrib/externs to `{*}`?

I can see that this tool is probably just used to prevent renaming, however, if your'e going to bother having a mode (the default mode!) that does output type information you should make sure it's correct at least.

**Comment by evmar** [comment]
I agree that we should make `--untyped` the default for now.  The eventual intent is to produce types and drop that flag, but for now the flag exists because there are bugs like the one you reported.

**Comment by evmar** [comment]
Filed #394 about the flag.

**Comment by tadeegan** [comment]
Is it that difficult to fix?

**Comment by evmar** [comment]
I don't think so.

**Comment by gkdn** [comment]
We tried running the tool for https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/react/index.d.ts and it is missing plenty of type information. There are missing @interface, template types and @extends ignored  etc.

Pretty much unusable in g3.

**Comment by evmar** [comment]
When you say "running the tool", do you mean generating externs from React's index.d.ts?
Could you open bugs on the issues you encountered?  This bug is about "declare namespace", so it's not helpful to bring up unrelated bugs.

Also, what do you mean by "in g3"?  If you mean google3, we already use tsickle for all TypeScript code... (?)

**Comment by gkdn** [comment]
> When you say "running the tool", do you mean generating externs from React's index.d.ts?

Yes.

> Could you open bugs on the issues you encountered? This bug is about "declare namespace", so it's not helpful to bring up unrelated bugs.

We hit the namespace issue as part of this so I thought there might be relation between the two.

> Also, what do you mean by "in g3"? If you mean google3, we already use tsickle for all TypeScript code...

Yes, g3 -> google3. Please use the tsickle on React.d.ts and you will see that is not usable without modifications.

**Comment by evmar** [comment]
For a bit more context, we have two sorts of milestones for our output.

The first is "make Closure Compiler accept, optimize, and not break TypeScript code".  For that we drop all the type annotations and do the other sorts of transformations described in https://github.com/angular/tsickle#what-conversion-means .  (Note that it's a lot more than just "prevent renaming".)  In #394 I made this behavior the default.

The second is "provide all types to allow type-based optimizations of output".  In our measurements this saves about 2% of app size and requires Closure and TypeScript to agree on types (which is especially hard in the presence of TS features not supported by Closure).  We're still working on it.

For the purposes of generating externs for use in a non-TypeScript project, you're on a use case we haven't really been working on but I think you really only care about the second level.  In particular we disallow "namespace" in google3 TypeScript code (both for style and codegen reasons) so we hadn't yet exercised this corner of the language.

Similarly we hadn't spent any time on template types because nobody has noticed them not working.  It would be worth filing a bug on that so it's not forgotten.

PS: I learned recently that templates aren't checked on Closure either.
Try compiling this and see that it has no errors.  (If you delete the first line it will find an error but not for the reasons you might expect...)
https://closure-compiler-debugger.appspot.com/#input0%3Dvar%2520x%2520%253D%2520%257Bboom%253A%25203%257D%253B%250A%250A%252F**%250A%2520*%2520%2540template%2520T%250A%2520*%2520%2540param%2520%257BT%257D%2520param%2520x%250A%2520*%252F%250Afunction%2520foo(param)%2520%257B%250A%2520%2520param.boom()%253B%250A%257D%250A%250Afoo('hello')%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26MISSING_PROPERTIES%3D1%26TRANSPILE%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1

**Comment by gkdn** [comment]
> I learned recently that templates aren't checked on Closure either.
> Try compiling this and see that it has no errors. (If you delete the first line it will find an error but not for the reasons you might expect...)

For completeness; this not true. Templates are [checked in JsCompiler](https://closure-compiler-debugger.appspot.com/#input0%3D%250A%252F**%250A%2520*%2520%2540template%2520T%250A%2520*%252F%250Aclass%2520X%2520%257B%250A%2520%2520%252F**%2520%2540return%2520%257BT%257D%2520*%252F%250A%2520%2520boom()%2520%257B%257D%250A%257D%250A%250A%252F**%2520%2540type%2520%257BX%253Cstring%253E%257D%2520*%252F%250Avar%2520x%253B%250A%250A%252F**%2520%2540type%2520%257BX%257D%2520*%252F%250Avar%2520y%2520%253D%2520x.boom()%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26MISSING_PROPERTIES%3D1%26TRANSPILE%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1). Old type inference is a different story where AFAIK templates were an after-thought and work poorly; while new type inference errors out in more cases you would expect.

**Comment by evmar** [comment]
Oh, I am sorry if I provided misinformation!  What did I do wrong with the example I gave?  (Comment out the first line and see that it compiles with no error.)  I don't get how your link relates to the one I gave?  Is it something about using a function vs a class?

**Comment by gkdn** [comment]
I was commenting on "templates aren't checked on Closure". They are actually checked but sometimes it is a bit loose in the old type inference - link in your example. If you switch to new type inference (NTI) you will see it catches more errors (if not saying all because NTI is not completed and might have rough edges).

(I used class just for the simpler syntax)
